src/main/java/net/neoforged/neoforge/transfer/resources/UnsafeResourceUtils.java

```
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.resources;
+
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.neoforge.fluids.FluidStack;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * <strong>Avoid use of this utility.</strong> This is only intended to be used internally by Neo or in read only
+ * scenarios to avoid allocation where you can ensure no data is written back to the stack.
+ */
+@ApiStatus.Internal
+public final class UnsafeResourceUtils {
```

Same with transactions, I really don't like these Unsafe classes.

What are the use cases for these in later slices?

---

So, in terms of read only access to an inner ItemStack or FluidStack, there is no way to do that safely. Instead of allocating every time you have a pure get method, these are provided. 

Let's take `WorldlyContainerWrapper` where it inquires `canTakeItemThroughFace`. There is no reason we should allocate here and instead just use the existing stack. 

The second scenario, is where something of a stack is a read only method, but the resource doesn't provide it. Instead of having to make a PR and wait, they can access it in a manner that expects them to be careful with its use.

---

The issue with that is noone will make a PR, they will just silently use the unsafe workaround forever. This has happened countless times over the lifetime of Forge.

Looking at `WorldlyContainerWrapper` it makes sense why this exists.

We should either declare getting the underlying stack part of public api with some similar warnings (potentially with some modification detection), or just clone the stack, I do not want 'internal' apis exposed or used by modders in any way.

---

Indeed, it should for sure not be `@ApiStatus.Internal` if it is meant for usage in mods.

In my version of the PR, the item resource was in the same package as the container wrapper, which allowed the access to the underlying stack to remain package-private.

---

From what I can find, `@Deprecated` is the only loosely fitting one besides `internal`. Would that be a better fit for it? While I understand the argument on "no one will make a PR for it if we add it", not doing this would not solve anything in that regard.

We have 3 scenarios: 
1. Ideal scenario: Someone sees a missing method and makes a PR.
2. Someone sees a missing method, they do `unsafe` access to avoid allocation in read only methods and likely doesn't make a PR (nor would I expect them to).
3. Someone sees a missing method and uses toStack() which allocates, but still no PR (But they should if using this way).

---

I suggest we just always clone, this is too much internal leakage, I _DO NOT_ want internal apis being used by modders in any situation.

ItemStack copies are not deep copies due to `PatchedDataComponentMap`, they should be very free to create.

---

Fabric never exposes publically the internal stack. I would always expose a copy to begin with, and consider exposing the underlying instance later if the copying is shown to be a bottleneck in real-world scenarios.